### ISSUE TITLE
Rho is calculated multiple times

### ISSUE THREAD

**Issue by rawbby** [lettucecfd/lettuce#90]
Rho is required for the calculation of equilibrium and u (and u_eq).
Therefor it should be pre calculated for both.

Atm:
```py
class lattices
    def u(self, f):
        return self.j(f) / self.rho(f)
```
```py
class BGKCollision:
    def __call__(self, f):
        rho = self.lattice.rho(f)
        u_eq = 0 if self.force is None else self.force.u_eq(f)
        u = self.lattice.u(f) + u_eq
        feq: torch.Tensor = self.lattice.equilibrium(rho, u)
        Si = 0 if self.force is None else self.force.source_term(u)
        return f - 1.0 / self.tau * (f - feq) + Si
```

My suggestion:
```py
class lattices
    def u(self, f, rho_):
        return self.j(f) / rho_
```
```py
class BGKCollision:
    def __call__(self, f):
        rho_ = self.lattice.rho(f)
        u_eq = 0 if self.force is None else self.force.u_eq(f)
        u = self.lattice.u(f, rho_) + u_eq
        feq: torch.Tensor = self.lattice.equilibrium(rho_, u)
        Si = 0 if self.force is None else self.force.source_term(u)
        return f - 1.0 / self.tau * (f - feq) + Si
```

**Comment by Olllom** [comment]
@rawbby A PR would be welcome.

One suggestion: make the `rho` argument for `u` optional (default to `None`); if it's not specified it will be computed in `u` as before.

```
class Lattice ...
    def u(self, f, rho=None):
        rho = self.rho(f) if rho is None else rho
        return self.j(f) / rho
```

This ensures backwards compatibility and keeps the API more user friendly, while giving you the same performance.

Also: In my opinion there is no need for typing (`feq: torch.Tensor ...`) or the `_`-suffix. Although it might be good practice, it makes it less readable and less consistent with the rest of the code. For local variables, I find typing to be completely unnecessary. For function parameters, it makes sense although we have consistently not used it so far (let's just keep it that way).

**Comment by rawbby** [comment]
Yeah, ```: torch.Tensor``` was just a note in my branch. The ```_```-suffix was because the name "```rho```" was already used for a method in the same namespace...

**Comment by Olllom** [comment]
ah, yeah, well. The namespace-issue is not really an issue in Python, so let's keep the name `rho`

**Comment by Olllom** [comment]
solved in #114
